lib/puppet/pops/validation/checker4_0.rb

```diff
@@ -535,18 +536,39 @@ def internal_check_future_reserved_word(o, name)
   NO_NAMESPACE = :no_namespace
   NO_PATH = :no_path
   BAD_MODULE_FILE = :bad_module_file
-  def internal_check_file_namespace(o, name, file)
+  def internal_check_file_namespace(o)
+    file = o.locator.file
     return if file.nil? || file == '' #e.g. puppet apply -e '...'
 
     file_namespace = namespace_for_file(file)
     return if file_namespace == NO_NAMESPACE
 
     # Downcasing here because check is case-insensitive
-    if file_namespace == BAD_MODULE_FILE || !name.downcase.start_with?(file_namespace)
-      acceptor.accept(Issues::ILLEGAL_DEFINITION_LOCATION, o, {:name => name, :file => file})
+    if file_namespace == BAD_MODULE_FILE || !o.name.downcase.start_with?(file_namespace)
+      acceptor.accept(Issues::ILLEGAL_DEFINITION_LOCATION, o, {:name => o.name, :file => file})
     end
   end
 
+  def internal_check_top_construct_in_module(prog)
+    return unless prog.is_a?(Model::Program) && !prog.body.nil?
+
+    #Check that this is a module autoloaded file
+    file = prog.locator.file
+    return if file.nil?
+    return if namespace_for_file(file) == NO_NAMESPACE
+
+    body = prog.body
+    return if (body.is_a?(Model::BlockExpression) && body.statements.all? { |s| valid_top_construct?(s) }) ||
+      valid_top_construct?(body)
+
+    name = body.respond_to?(:name) ? "'#{body.name}' " : nil
+    acceptor.accept(Issues::ILLEGAL_TOP_CONSTRUCT_LOCATION, body, {:name => name, :file => file})
```

Comment:
This will only give you one warning even if there are multiple illegal top constructs in the file, and it will associate it with the source location of the body (which will always be line 1).
You could map the body to an array of invalid top constructs, and return if it is empty, otherwise accept an issue for each of the invalid constructs. This way it is possible to collect all of the issues
with correct source location information.